### ISSUE TITLE
MPS device neural net in validation gives the same results after n-th batch element

### ISSUE THREAD

**Issue by hajdurobert92** [pytorch/pytorch#80844]
### 🐛 Describe the bug

So the problem is basically, that on MPS device, at validation phase, neural net gives the same result/logits after 4-5th element in a batch. Input vector is different. It works normally on cpu device.

Here's an example code:

```

#imports
import torch
import torchvision
import numpy as np
import matplotlib.pyplot as plt
import torch.nn as nn
from torch import optim
from torchvision.transforms import transforms
import torch.nn.functional as F
device = torch.device('mps')


# COMMAND ---------

#download the dataset(s)
train_dataset = torchvision.datasets.MNIST('/Users/xy/work/content',train=True,
                                          transform=transforms.ToTensor(),download=True)

test_dataset = torchvision.datasets.MNIST('/Users/xy/work/content',train=False,
                                          transform=transforms.ToTensor(),download=True)

# COMMAND ----------

#create dataloaders from datasets
train_dataloader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                  batch_size=64,shuffle=True)

test_dataloader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                  batch_size=64,shuffle=True)

# COMMAND ----------

#plot first 10 images
images, labels = next(iter(train_dataloader))


#plt.figure(figsize=(10,8))
#for i in range(10):
  #  plt.subplot(2,5,i+1)
 #   plt.imshow(images[i][0],cmap='BuPu')
#plt.show()

# COMMAND ----------

#our first neural network
class MNIST_model(nn.Module):
    def __init__(self):
        super(MNIST_model,self).__init__()
        self.hidden_layer_1=nn.Linear(28*28,128)
        self.hidden_layer_2=nn.Linear(128,64)
        self.output_layer=nn.Linear(64,10)
    def forward(self,x):
        x = F.relu(self.hidden_layer_1(x)) 
        x = F.relu(self.hidden_layer_2(x))
        x = self.output_layer(x)
        return x

# COMMAND ----------

#instantiate the model and print paramters
model=MNIST_model()
model.to(device)

print(model.parameters)

# COMMAND ----------

#print dataset shape, reshape dataset to the desired format
print(images.shape)
images = images.view(images.shape[0], -1)
print(images.shape)

# COMMAND ----------

#pass images (data) trough our model:
logps = model(images.to(device)) #log probabilities
print(logps.shape) # print output shape
print(logps) # print output

# COMMAND ----------

#create a loss function and calculate a loss
criterion = nn.CrossEntropyLoss()
loss = criterion(logps.detach().cpu(), labels) #calculate the NLL loss

# COMMAND ----------

#training loop
optimizer=torch.optim.Adam(model.parameters(),lr=0.01)
epochs = 15
for e in range(epochs):
    running_loss = 0
    for images, labels in train_dataloader:
        # Flatten MNIST images into a 784 long vector
        images = images.view(images.shape[0], -1)
    
        # Training pass
        optimizer.zero_grad()
        
        logits = model(images.to(device))
        #print(logits)
        loss = criterion(logits, labels.to(device))
        
        #This is where the model learns by backpropagating
        loss.backward()
        
        #And optimizes its weights here
        optimizer.step()
        
        running_loss += loss.item()
    else:
        print("Epoch {} - Training loss: {}".format(e + 1, running_loss/len(train_dataloader)))


# COMMAND ----------

#validate our model
correct_count, all_count = 0, 0
for images,labels in test_dataloader:
  for i in range(len(labels)):
    img = images[i].view(1, 784)
    print(img)
    with torch.no_grad():
        logps = model(img.to(device))

    #print(logps)
    ps = torch.exp(logps)
    probab = list(ps.detach().cpu().numpy()[0])
    pred_label = probab.index(max(probab))
    true_label = labels.numpy()[i]
    #print(str(pred_label) + " " + str(true_label))
    if(true_label == pred_label):
      correct_count += 1
    all_count += 1

print("Number Of Images Tested =", all_count)
print("\nModel Accuracy =", (correct_count/all_count))

# COMMAND ----------

import numpy as np
def view_classify(img, ps):
    ''' Function for viewing an image and it's predicted classes.
    '''
    ps = ps.squeeze()

    fig, (ax1, ax2) = plt.subplots(figsize=(6,9), ncols=2)
    ax1.imshow(img.resize_(1, 28, 28).numpy().squeeze())
    ax1.axis('off')
    ax2.barh(np.arange(10), ps)
    ax2.set_aspect(0.1)
    ax2.set_yticks(np.arange(10))
    ax2.set_yticklabels(np.arange(10))
    ax2.set_title('Class Probability')
    ax2.set_xlim(0, 1.1)
    plt.tight_layout()

images, labels = next(iter(test_dataloader))

img = images[0].view(1, 784)
with torch.no_grad():
    logps = model(img.to(device))

    
ps = torch.exp(logps)
probab = list(ps.detach().cpu().numpy()[0])
data = probab
norm = (data - np.min(data)) / (np.max(data) - np.min(data))

print("Predicted Digit =", probab.index(max(probab)))
view_classify(img.view(1, 28, 28), norm)

# COMMAND ----------

```

### Versions

Collecting environment information...
PyTorch version: 1.13.0.dev20220704
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: version 3.20.1
Libc version: N/A

Python version: 3.10.5 (v3.10.5:f377153967, Jun  6 2022, 12:36:10) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] torch==1.13.0.dev20220704
[pip3] torchaudio==0.14.0.dev20220603
[pip3] torchvision==0.14.0.dev20220704
[conda] numpy                     1.20.2           py38hbf7bb01_0    conda-forge

cc @kulinseth @albanD

**Comment by qqaatw** [comment]
Hi, can you help narrow down the scope and provide a minimal repro that manifests the issue?  For example, if the training part is not needed in order to reproduce the issue, you can remove it. Thanks.

**Comment by hajdurobert92** [comment]
Hi!

This is the minimum reproduction code. You can run the train and with the same model the error occurs at validation.

Thank you

**Comment by DenisVieriu97** [comment]
Hi @hajdurobert92! Thanks for the report - I've put a fix for this issue at https://github.com/kulinseth/pytorch/pull/64. I will update back this thread when the PR gets into PyTorch master.

**Comment by DenisVieriu97** [comment]
This is fixed in the latest PyTorch nightly
